### PR TITLE
[feature] Refactor and optimize local buffer concept and optimize rotations

### LINKED ISSUE #149: Writeout: assess if the time window where capture is blocked can be reduced further
Observed on 77c4f356 (high throughput host): drops right after rotation of map. This is probably due to the fact that the capture is blocked and the buffer runs full before capture is continued.
```
╭──────────────────────────────────────────────────────────────────────────────╮
│                              Interface Statuses                              │
├───────┬──────────┬────────────┬───────────┬─────────────┬───────────┬────────┤
│       │    total │            │     total │             │           │ active │
│ iface │ received │ + received │ processed │ + processed │ + dropped │ for    │
├───────┼──────────┼────────────┼───────────┼─────────────┼───────────┼────────┤
│ eth40 │   4.74 k │     7.00   │    4.74 k │      4.00   │         0 │ 18m30s │
│ eth43 │   8.63 M │     9.73 k │    8.62 M │     10.19 k │         0 │ 18m30s │
│ eth60 │  99.14 M │    72.28 k │   99.14 M │     91.54 k │         0 │ 18m30s │
│ eth64 │ 341.10 M │   256.73 k │  340.92 M │    282.87 k │       399 │ 18m30s │
╰───────┴──────────┴────────────┴───────────┴─────────────┴───────────┴────────╯

Runtime info:

            Running since: 2023-07-31 16:56:30 (18m30s ago)
  Last scheduled writeout: 2023-07-31 17:15:00 (1s ago)

Totals:

    Packets
       Received: 448.87 M / + 338.75 k
      Processed: 448.69 M / + 384.60 k
        Dropped:      + 399
```
DoD:

- [x] check if there's a way to have a temporary buffer while rotating
- [x] @els0r Run test on high-profile host(s)
- [x] Clean up code and optimize performance (currently very crude)

### PR DESCRIPTION
**Pull request by fako1024**
@els0r This should bring it to the next (and hopefully last) level for the release. I've given the whole concept some more thought and think this should get rid of all issues. What has been done (see comments in code for details):

- Some micro-optimizations in the rotation / transfer logic (especially for existing flows that get removed due to inactivity).
- Instead of using an unlimited shared memory pool the CaptureManager now uses a single _limited_ memory pool (which in practice probably can use a singular element) that is governed by a global config (instead of per interface, which should make life a lot easier) and puts a global limit to the maximum memory consumption.
- Rotations are now performed sequentially rather than in a RunGroup (which in practice was the case already due to the sequential nature of the writeout to disk). This not only allows us to use a single buffer (see previous point) but also reduces congestion.
- The way the shared buffer is now handled will minimize the interface "downtime" during rotation because the lock is only established once we _know_ that the memory buffer is available / drained.
- The interface status obtained during rotations is now fetched in the background (while the interface is locked) to further minimize the "downtime".

Closes #149

### CODE REVIEW

File: pkg/capture/capture.go
Comment (fako1024) [inline pkg/capture/capture.go:222]: Size limit is now governed globally (via `setLocalBuffers()`, respectively by reasonable defaults)

File: pkg/capture/capture.go
Comment (fako1024) [inline pkg/capture/capture.go:385]: Getting the status can take a moment, but is independent from the actual rotation (since the capture is locked there is no potential race with the counters), so we can do it in the background _while_ doing the rotation. Depending on how long it actually takes this can save some time during which buffering is required.

File: pkg/capture/capture_manager.go
Comment (fako1024) [inline pkg/capture/capture_manager.go:416]: See comment. Important: The whole concept is still concurrency-safe (I tested it, the memory pool ensure that no race conditions can occur), I just _choose_ to not do it in parallel because there's no benefit (on the contrary). External status calls are still done in parallel (however in practice are run one by one since the default number of buffers is one). IMHO this should be the optimal use, but it still allows us to pimp for edge cases (we have to test if maybe calls to `Status()` for all interfaces take too long if many interfaces are present).

File: pkg/capture/capture_test.go
Comment (fako1024) [inline pkg/capture/capture_test.go:60]: One nice side effect: Due to all the optimizations I was able to bump this back up to 1000 iterations again (which I take as a good sign)!

File: pkg/capture/flow.go
Comment (fako1024) [inline pkg/capture/flow.go:234]: This was moved inside the condition below (because it is not required for flows that get dropped).

File: pkg/capture/flow.go
Comment (fako1024) [inline pkg/capture/flow.go:238]: I removed the `HasBeenIdle()` because I saw that for some reason the compiler did not inline this trivial check. Since it is only used once (here) and I added a comment I think that's acceptable.

File: pkg/capture/capture_manager.go
Comment (fako1024) [inline pkg/capture/capture_manager.go:440]: @els0r Here I'd like your opinion: I'd really like to have a way to get an idea of the time an interface is locked (not the global rotation time). From my point of view this is easiest with such a debug log statement, but metrics would be more pretty of course. However, I recall you saying that cardinality might be a problem, hence I'm not sure if that's a viable option (because of course we need to know which interface is affected). What's your take on this (keeping in mind that it might also be feasible to drop this entirely in case we don't even see any drops anymore :stuck_out_tongue: ).

File: pkg/capture/capture.go
Comment (fako1024) [inline pkg/capture/capture.go:431]: This little one actually gave me some headache (and took the longest to figure out, ironically). I always ran into deadlocks at first due to the `done` channel now also having a depth of one _and_ me trying to fetch the buffer element at the beginning of `lock()`. Since it was possible that a concurrent `status()` call (man am I glad I added these concurrency sledgehammer tests) was already queuing up _while_ the capture was in buffer mode _and_ blocked in the `PPOLL()` it was possible that everything got stuck.
This additional `Unblock()` will basically ensure that in such scenarios the number of calls is "balanced" again and there's always an unblock event to save the day... :rofl:

File: pkg/capture/capture_manager.go
Comment (els0r) [reply pkg/capture/capture_manager.go:440]: While metrics would be optimal in order to continuously measure this per interface, I would also go with a simple log line here. It's after all something you want to investigate ad-hoc rather than baseline for tool health.

Comment (fako1024) [reply pkg/capture/capture_manager.go:440]: Sounds good to me, thx for the feedback!

**Review by els0r — APPROVED** [review]
Will test this on sensor fleet after merge.

**Pull request merged**
